felix/wireguard/bootstrap.go

```diff
@@ -17,110 +17,401 @@ package wireguard
+		logCxt.Info("No valid wireguard kernel routing - removing wireguard configuration completely")
+		return nil, removeWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	storedPublicKey, err := getPublicKeyForNode(logCxt, nodeName, calicoClient)
+	if err != nil {
+		// Could not determine the public key for this node.
+		return nil, fmt.Errorf("couldn't determine current wireguard configuration after %d retries: %v", bootstrapMaxRetries, err)
+	}
+
+	if storedPublicKey != kernelPublicKey {
+		// The public key configured in the kernel differs from the value stored in the node. Remove all wireguard
+		// configuration.
+		logCxt.Info("Found mismatch between kernel and datastore wireguard keys - removing wireguard configuration")
+		return nil, removeWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	// The configured and stored wireguard key match.
+	logCxt.WithField("peerKeys", kernelPeerKeys).Info("Wireguard public key matches kernel")
+	return kernelPeerKeys, nil
+}
+
+// FilterTyphaEndpoints filters the supplied set of typha endpoints to the set where wireguard routing is most likely
+// to succeed. Any errors encountered are swallowed and the associated peer is just included.
+func FilterTyphaEndpoints(
+	configParams *config.Config,
+	v3Client clientv3.Interface,
+	typhas []discovery.Typha,
+	peersToValidate set.Set,
+) []discovery.Typha {
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
+		// This should not change for a given deployment. Only host encryption should impact typha connectivity.
+		log.Debug("No host encryption - all typhas should be accessible")
+		return typhas
+	}
+	if peersToValidate == nil {
+		// Peers to validate will be nil if there are no wireguard peers in the kernel that we need to check.
+		log.Debug("Wireguard is not configured locally")
+		return typhas
+	}
+
+	log.Debugf("Filtering typha endpoints for wireguard: %v", typhas)
+
+	var filtered []discovery.Typha
+
+	for _, typha := range typhas {
+		logCxt := log.WithField("typhaAddr", typha.Addr)
+		if typha.NodeName == nil {
+			logCxt.Info("Typha endpoint has no node information - include typha endpoint")
+			filtered = append(filtered, typha)
+			continue
+		}
+
+		typhaNodeName := *typha.NodeName
+		logCxt = logCxt.WithField("typhaNodeName", typhaNodeName)
+		if typhaNodeName == configParams.FelixHostname {
+			// This is a local typha. We should always be able to connect.
+			logCxt.Info("Typha endpoint is local - include typha endpoint")
+			filtered = append(filtered, typha)
+			continue
+		}
+
+		// Get the public key configured for the typha node.
+		typhaNodeKey, err := getPublicKeyForNode(logCxt, typhaNodeName, v3Client)
+		if err != nil {
+			// If we were unable to determine the public key then just include the endpoint.
```

Comment:
What scenarios would we hit an error getting the public key for a remote node? Do we want to filter the endpoint out in that case?

Reply:
My take is including the nodes is the safer thing to do - it just slows down how quickly we resume connection to typha.

Reply:
I think this reasoning as to why we are including it would be nice to be added as a comment here.